xplique/example_based/semifactuals.py

```diff
@@ -0,0 +1,233 @@
+        `tf.data.Dataset` are assumed to be batched as tensorflow provide no method to verify it.
+        Batch size and cardinality of other datasets should match `cases_dataset`.
+        Be careful, `tf.data.Dataset` are often reshuffled at each iteration, be sure that it is not
+        the case for your dataset, otherwise, examples will not make sense.
+    k
+        The number of examples to retrieve per input.
+    projection
+        Projection or Callable that project samples from the input space to the search space.
+        The search space should be a space where distances are relevant for the model.
+        It should not be `None`, otherwise, the model is not involved thus not explained.
+
+        Example of Callable:
+        ```
+        def custom_projection(inputs: tf.Tensor, np.ndarray, targets: tf.Tensor, np.ndarray = None):
+            '''
+            Example of projection,
+            inputs are the elements to project.
+            targets are optional parameters to orientated the projection.
+            '''
+            projected_inputs = # do some magic on inputs, it should use the model.
+            return projected_inputs
+        ```
+    case_returns
+        String or list of string with the elements to return in `self.explain()`.
+        See the base class returns property for more details.
+    batch_size
+        Number of sample treated simultaneously for projection and search.
+        Ignored if `tf.data.Dataset` are provided (those are supposed to be batched).
+    distance
+        Distance for the FilterKNN search method.
+        Distance function for examples search. It can be an integer, a string in
+        {"manhattan", "euclidean", "cosine", "chebyshev", "inf"}, or a Callable,
+        by default "euclidean".
+    """
+    # pylint: disable=duplicate-code
+
+    _returns_possibilities = [
+        "examples", "weights", "distances", "labels", "include_inputs",
+        "nuns", "nuns_indices", "dist_to_nuns", "nuns_labels"
+    ]
+
+    def __init__(
+        self,
+        cases_dataset: Union[tf.data.Dataset, tf.Tensor, np.ndarray],
+        targets_dataset: Union[tf.data.Dataset, tf.Tensor, np.ndarray],
+        labels_dataset: Optional[Union[tf.data.Dataset, tf.Tensor, np.ndarray]] = None,
+        k: int = 1,
+        projection: Union[Projection, Callable] = None,
+        case_returns: Union[List[str], str] = "examples",
+        batch_size: Optional[int] = 32,
+        distance: Union[int, str, Callable] = "euclidean",
+    ):
+
+        super().__init__(
+            cases_dataset=cases_dataset,
+            labels_dataset=labels_dataset,
+            targets_dataset=targets_dataset,
+            k=k,
+            projection=projection,
+            case_returns=case_returns,
+            batch_size=batch_size,
+        )
+
+        # set distance function and order for the search method
+        self.distance = distance
+        self.order = ORDER.ASCENDING
```

Comment:
The distance and the order do not seem to be used in this class. I think they can be removed from the object.

It might also be the case for similar examples and counterfactuals.